test/rulesets/Base/array.jl

```diff
@@ -101,7 +101,51 @@ end
     test_rrule(hvcat, 1, rand(3)', transpose(rand(3)) ⊢ rand(1,3); check_inferred=VERSION>v"1.1")
 end
 
+@testset "reverse" begin
+    # Forward
+    test_frule(reverse, rand(5))
+    test_frule(reverse, rand(5), 2, 4)
+    test_frule(reverse, rand(5), fkwargs=(dims=1,))
+
+    test_frule(reverse, rand(3,4), fkwargs=(dims=2,))
+    if VERSION >= v"1.6"
+        test_frule(reverse, rand(3,4))
+        test_frule(reverse, rand(3,4,5), fkwargs=(dims=(1,3),))
+    end
+
+    # Reverse
+    test_rrule(reverse, rand(5))
+    test_rrule(reverse, rand(5), 2, 4)
+    test_rrule(reverse, rand(5), fkwargs=(dims=1,))
+
+    test_rrule(reverse, rand(3,4), fkwargs=(dims=2,))
+    if VERSION >= v"1.6"
+        test_rrule(reverse, rand(3,4))
+        test_rrule(reverse, rand(3,4,5), fkwargs=(dims=(1,3),))
+
+        # Structured
+        @test unthunk(rrule(reverse, Diagonal([1,2,3]))[2](rand(3,3))[2]) isa Diagonal
+    end
+end
+
+@testset "circshift" begin
+    # Forward
+    test_frule(circshift, rand(10), 1)
+    test_frule(circshift, rand(10), (1,) ⊢ NoTangent())  # surely it's a bug that omitting this gives InexactError: Int64(0.99) and many screens of [34] _make_j′vp_call(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, ...
```

Comment:
for me
```julia
julia> rand_tangent((2,))
NoTangent()
```
Which means it should work without the explicit tangent. Are you seeing different behaviour?

Reply:
I was getting:
```
julia> test_frule(circshift, rand(10), (1,) ⊢ NoTangent())
Test Summary:                                  | Pass  Total
test_frule: circshift on Vector{Float64},Int64 |    4      4
Test.DefaultTestSet("test_frule: circshift on Vector{Float64},Int64", Any[], 4, false, false)

julia> test_frule(circshift, rand(10), (1,))
test_frule: circshift on Vector{Float64},Int64: Error During Test at /Users/me/.julia/packages/ChainRulesTestUtils/AX7fv/src/testers.jl:122
  Got exception outside of a @test
  DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 11 and 10")
  Stacktrace:
    [1] _bcs1
      @ ./broadcast.jl:516 [inlined]
    [2] _bcs
      @ ./broadcast.jl:510 [inlined]
    [3] broadcast_shape
      @ ./broadcast.jl:504 [inlined]
    [4] combine_axes
      @ ./broadcast.jl:499 [inlined]
    [5] instantiate
      @ ./broadcast.jl:281 [inlined]
    [6] materialize
      @ ./broadcast.jl:904 [inlined]
    [7] (::FiniteDifferences.var"#73#74"{FiniteDifferences.var"#75#76"{FiniteDifferences.var"#77#78"{ChainRulesTestUtils.var"#fnew#38"{ChainRulesTestUtils.var"#call_on_copy#42"{NamedTuple{(), Tuple{}}}, Tuple{typeof(circshift), Vector{Float64}, Tuple{Int64}}, Tuple{Bool, Bool, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64}, Tuple{Int64}, Tuple{FiniteDifferences.var"#Real_from_vec#20"}}}}}, Vector{Float64}, Vector{Float64}})(ε::Float64)
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/W3rQO/src/grad.jl:46
    [8] #10
      @ ./broadcast.jl:329 [inlined]
    [9] macro expansion
      @ ~/.julia/packages/StaticArrays/gBeIF/src/broadcast.jl:126 [inlined]
   [10] _broadcast
      @ ~/.julia/packages/StaticArrays/gBeIF/src/broadcast.jl:100 [inlined]
   [11] copy
      @ ~/.julia/packages/StaticArrays/gBeIF/src/broadcast.jl:27 [inlined]
   [12] materialize
      @ ./broadcast.jl:904 [inlined]
   [13] _eval_function(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#73#74"{FiniteDifferences.var"#75#76"{FiniteDifferences.var"#77#78"{ChainRulesTestUtils.var"#fnew#38"{ChainRulesTestUtils.var"#call_on_copy#42"{NamedTuple{(), Tuple{}}}, Tuple{typeof(circshift), Vector{Float64}, Tuple{Int64}}, Tuple{Bool, Bool, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64}, Tuple{Int64}, Tuple{FiniteDifferences.var"#Real_from_vec#20"}}}}}, Vector{Float64}, Vector{Float64}}, x::Float64, step::Float64)
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/W3rQO/src/methods.jl:249
   [14] _estimate_magnitudes(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#73#74"{FiniteDifferences.var"#75#76"{FiniteDifferences.var"#77#78"{ChainRulesTestUtils.var"#fnew#38"{ChainRulesTestUtils.var"#call_on_copy#42"{NamedTuple{(), Tuple{}}}, Tuple{typeof(circshift), Vector{Float64}, Tuple{Int64}}, Tuple{Bool, Bool, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64}, Tuple{Int64}, Tuple{FiniteDifferences.var"#Real_from_vec#20"}}}}}, Vector{Float64}, Vector{Float64}}, x::Float64)
```
But if I make a new session and get Ch.T.U. 1.0, then it does work. (Looks like this was ChainRulesTestUtils v0.7.13 via fall-through, locally.)

However, CI also gave me that error, which is where I coped it from. We can try again and see if it's changed its mind. Surely the bounds have not changed during this PR though. It's a bit weird.

Reply:
oh ok, interesting about the CI - I think CRTU 1.0 solves a few annoying things yeah. looking at the commented function now

Reply:
Seems to work now!